do not reload the page if the server is (temporarily) unreachable

ChristophWurst opened the pull request:
Since it's quite annoying to see an ugly browser error message like `Server not found` when resuming my notebook, it would be great if we would _not_ reload in that case.
- [x] TODO: unit tests

cc @jancborchardt since we talked about that

ChristophWurst committed:
do not reload the page if the server is (temporarily) unreachable

MorrisJobke commented:
> TODO: unit tests

@ChristophWurst ping ;)

ChristophWurst commented:
🏓 pong

LukasReschke commented:
🏓 ping

MorrisJobke commented:
I don't really like this, because with this then you can navigate and it always show empty panes. It seems to work from an end user perspective. Maybe add the notification back to give some user feedback.

ChristophWurst commented:
@MorrisJobke @jancborchardt what should it look like? Show a permanent notification if the connection was lost? Show a temporary connection? Or something different?

MorrisJobke commented:
> @MorrisJobke @jancborchardt what should it look like? Show a permanent notification if the connection was lost? Show a temporary connection? Or something different?

 would say a temporary notification if you browse and there is no return (no connection available).

ChristophWurst commented:
@MorrisJobke I updated my commit and added a throttled temporary notification. I've also added a simple test for it. However, I could not test the throttling because underscore doesn't use ``setTimeout`` internally, therefore I'm unable to fake the timer.

These are the failing tests:
```js
		it('doesn\'t show more than one connection lost notification at once', function() {
			var xhr = { status: 0 };
			spyOn(OC.Notification, 'showTemporary');

			$(document).trigger(new $.Event('ajaxError'), xhr);
			clock.tick(100);
			$(document).trigger(new $.Event('ajaxError'), xhr);
			clock.tick(100);
			$(document).trigger(new $.Event('ajaxError'), xhr);

			expect(OC.Notification.showTemporary.calls.count()).toBe(1);
		});
		it('shows a new notification once the old one was removed', function() {
			var xhr = { status: 0 };
			spyOn(OC.Notification, 'showTemporary');

			$(document).trigger(new $.Event('ajaxError'), xhr);
			expect(OC.Notification.showTemporary.calls.count()).toBe(1);

			clock.tick(7200);
			$(document).trigger(new $.Event('ajaxError'), xhr);
			
			expect(OC.Notification.showTemporary.calls.count()).toBe(2);
		});
```
But we actually do not have to test the throttling, because we can rely on the 3rd party lib to handle that properly 🙈 

I've found a little concurrency problem in the notification code: when we fade out a single notification, we do not check if another one was added/removed meanwhile the 400ms delay. I can try to fix that in a separate PR if we consider this is a critical bug (important notifications might not be shown).
ref https://github.com/nextcloud/server/blob/07a1be2e06361d5c88eca53acb790e1dc499bdb1/core/js/js.js#L1057-L1069

nickvergessen commented:
Of course when you do a hard refresh (app change) it still shows the browser page, but at least after a resume (browser faster then network connector) it does not refresh now.

👍

MorrisJobke commented:
Tested and works :+1: 

@LukasReschke is it okay to not have the "signed-off" stuff? or is this required?

MorrisJobke commented:
> @LukasReschke is it okay to not have the "signed-off" stuff? or is this required?

@ChristophWurst Could you add the signed-off comment - see https://github.com/nextcloud/server/blob/master/CONTRIBUTING.md#sign-your-work
